mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp

```diff
@@ -306,6 +323,128 @@ struct ConvertingCopyKernelRewritePattern final
   }
 };
 
+/// Layout normalization.
+
+/// In most cases, we can treat the CYX or YXC parts of the filter dimension
+/// as a single unit for the purposes of vectorization, coordinate computations,
+/// and so on. However, if the filter layout is created by transposing the
+/// underlying memory in order to match another tensor, we cannot do this
+/// and so must flag when we have performed such a relayout.
+///
+/// This is a nominally temporary hack that should be removed once we can more
+/// precisely handle Unfold{} by static analysis instead of by heuristic.
+/// As such, it'll probably presist for quite some time.
+constexpr llvm::StringLiteral dontUnfoldAttrName("has_relayout_do_not_unfold");
+/// Make the dimensions that are the values in `mapping` and exist within
+/// `toLayout` be in the same relative order as the dimensions that the keys of
+/// `mappping` have within `fromLayout`, where both layout are given by the
+/// names of the attributes containing them.
+///
+/// To enable usage in rewrite patterns, returns failure() when no change is
+/// made.
+LogicalResult makeToLayoutLikeFromLayoutAlong(
+    PatternRewriter &b, RockConvInterface op, StringRef fromLayoutAttrName,
+    TypedValue<ShapedType> toArg, StringRef toLayoutAttrName,
+    const llvm::StringMap<StringAttr> &mapping) {
+  llvm::SmallVector<StringAttr> expectedOrder;
+  auto fromLayout = op->getAttrOfType<ArrayAttr>(fromLayoutAttrName);
+  auto toLayout = op->getAttrOfType<ArrayAttr>(toLayoutAttrName);
+  for (StringRef fromName : fromLayout.getAsValueRange<StringAttr>()) {
+    auto maybeCorresponding = mapping.find(fromName);
+    if (maybeCorresponding != mapping.end())
+      expectedOrder.push_back(maybeCorresponding->getValue());
+  }
+
+  llvm::SmallDenseMap<StringAttr, size_t> toLayoutIdxs;
+  for (auto pair : llvm::enumerate(toLayout.getAsRange<StringAttr>()))
+    toLayoutIdxs.insert({pair.value(), pair.index()});
+
+  bool inOrder = true;
+  size_t prevIndex = 0;
+  for (StringAttr expected : expectedOrder) {
+    size_t thisIndex = toLayoutIdxs.find(expected)->getSecond();
+    if (thisIndex <
+        prevIndex) { // the values are not in the relative expected order
+      inOrder = false;
+    }
+    // Defensively prevent possible weirdness that might arise when
+    // thisIndex < prevIndex.
+    prevIndex = std::max(thisIndex, prevIndex);
+  }
+  if (inOrder)
+    return failure();
+
```

Comment:
Nit: If I'm to write this, I prefer to just get rid of this first loop but instead:
 - Use transform to convert the `toLayout` to `newToLayout`
   - Use index as internal lambda implementation details to the transform
 - Then compare whether `toLayout` is equivalent to `newToLayout` to decide whether it is inOrder.

Reply:
But yeah, making a transform unconditionally and seeing if it's the identity is also reasonable ... but a lot more work most of the time?